### ISSUE TITLE
Deep linking doesn't work properly with StackNavigator

### ISSUE THREAD

**Issue by Garkavenko** [react-navigation/react-navigation#7908]
**Current Behavior**

When app was opened and link was pressed, react-navigation ignores 'initialRouteName' prop, so i cant go back from this screen. When app was closed, initialRouteName works fine and i can go back from the navigated screen.

**Expected Behavior**

Deep linking works in the same way when app is closed and opened

**How to reproduce**
repo: https://github.com/Garkavenko/react-navigation-deeplinks-snak (file App.js)

I have bottom tabs navigator with stack navigator inside each tab.
render:
```js
function First() {
  return (
    <View style={{ flex: 1 }}>
      <Text>First</Text>
    </View>
  )
}

function SubFirst() {
  return (
    <View style={{ flex: 1 }}>
      <Text>SubFirst</Text>
    </View>
  )
}

function SubSecond() {
  return (
    <View style={{ flex: 1 }}>
      <Text>SubSecond</Text>
    </View>
  )
}

function Second() {
  return (
    <View style={{ flex: 1 }}>
      <Text>Second</Text>
    </View>
  )
}

const Tab = createBottomTabNavigator();
const FirstStack = createStackNavigator();
const SecondStack = createStackNavigator();

function FistStackView() {
  return (
    <FirstStack.Navigator>
      <FirstStack.Screen name="first" component={First} />
      <FirstStack.Screen name="subfirst" component={SubFirst} />
    </FirstStack.Navigator>
  )
}

function SecondStackView() {
  return (
    <SecondStack.Navigator initialRouteName="second">
      <SecondStack.Screen name="second" component={Second}/>
      <SecondStack.Screen name="subsecond" component={SubSecond}/>
    </SecondStack.Navigator>
  )
}

const App = () => {
  const ref = useRef(null);

  const { getInitialState } = useLinking(ref, {
    prefixes: ['myapp://'],
    config: {
      secondTab: {
        initialRouteName: 'second',
        screens: {
          subsecond: 'test/:someParam',
          second: 'test',
        }
      },
    },
  });

  const [isReady, setIsReady] = React.useState(false);
  const [initialState, setInitialState] = React.useState();

  React.useEffect(() => {
    Promise.race([
      getInitialState(),
      new Promise(resolve =>
        // Timeout in 150ms if `getInitialState` doesn't resolve
        // Workaround for https://github.com/facebook/react-native/issues/25675
        setTimeout(resolve, 150)
      ),
    ])
      .catch(e => {
        console.error(e);
      })
      .then(state => {
        if (state !== undefined) {
          setInitialState(state);
        }
        setIsReady(true);
      });
  }, [getInitialState]);

  if (!isReady) {
    return null;
  }

  return (
    <>
      <StatusBar barStyle="dark-content" />
      <NavigationContainer ref={ref} initialState={initialState}>
        <Tab.Navigator>
          <Tab.Screen name="firstTab" component={FistStackView} />
          <Tab.Screen name="secondTab" component={SecondStackView} />
        </Tab.Navigator>
      </NavigationContainer>
    </>
  );
};
```
when i open link myapp://test/123 when app was closed, it works fine(initialRouteName works) and shows subsecond screen and back arrow, so i can go back to 'second' screen.
<img src="https://user-images.githubusercontent.com/62869905/77881614-06889100-7268-11ea-98ab-e650ac8e0722.jpg" width="200">

but when app was opened and second tab wasn't opened(lazy mode bottom tabs), so secondStack wasn't mounted at all, its working strange.
So, open the app, hide it and open myapp://test/123
It will show subsecond screen, but without back arrow, so i cant go back to 'second' screen
<img src="https://user-images.githubusercontent.com/62869905/77881897-8f9fc800-7268-11ea-893c-43c12b72fdbe.jpg" width="200">

**Your Environment**

| software                       | version |
| ------------------------------ | ------- |
| iOS or Android                 |
| @react-navigation/native       |^5.1.3
| @react-navigation/stack        |^5.2.8
| @react-navigation/bottom-tabs       |^5.2.4
| react-native-gesture-handler   |^1.6.1
| react-native-safe-area-context |^0.7.3
| react-native-screens           |^2.4.0
| react-native                   |0.61.2
| expo                           |
| node                           |v12.11.1
| npm or yarn                    |6.13.7

**Comment by compojoom** [comment]
I think that I'm running into the same issue. And that is somehow related to https://github.com/react-navigation/react-navigation/issues/7698

initialRouteName appears to be there only to specify a default Scene when you navigate to this Navigator. If you have a call that is specifying a different screen the initialRoute is not being rendered. 

I think that specifying lazy=false just makes it so that the initialRouteName is always rendered and when you add a new screen on the stack that new screen can go somewhere back.

**Comment by Garkavenko** [comment]
@compojoom yep, lazy=false could be a resolve, but in a real project i have some requests on the each tab when its mounted and render FlatList with complex items. So app will lags on start if every tabs will be mounted

**Comment by compojoom** [comment]
Yes, I know. I don't think that it's really the solution. Not sure what is supposed to happen. I think that if you navigate to a new tab on non-default tab there should be a back button to the initialRoute. At least this is how I imagine it. Right now there is no way to go back to the initial route. goBack navigates back to the previous tab, but it doesn't reset the tab you was on. pop - just triggers an error and same goes for reset. 

I'm not sure what the supposed flow was here. It seems like a simple and common thing?

**Comment by Garkavenko** [comment]
@compojoom, i dont think, that it is the common thing

It defenetly is a bug, because same deeplinking logic works in different ways because of this

And the main problem is the different working of 'initialRouteName' prop in the Stack.Navigator component and deep-linking config. I think, initialRouteName of Stack.Navigator prop should works the same as it works in deep-linking config(it should be always mounted as first screen in stack if exists in navigator), and it will resolve the most of problems with it. And in Stack.Navigator component should be 'defaultRouteName', which should works like initialRouteName works now(something like that).

**Comment by compojoom** [comment]
Very interesting. I just tried your code on iOS (had to add myapp as domain in xCode). When the app is closed and I open the deeplink it really behaves as you say - the initialRoute is rendered and there is a back button. 
When the app was started - opening the url was not changing the screens. (maybe this is an iOS problem, no idea).

I still continue to think that we have an error in the way we handle nested Navigators and that manifests in your deep-link problem. The docs here: https://reactnavigation.org/docs/nesting-navigators#navigating-to-a-screen-in-a-nested-navigator state that if you want to navigate to a nested navigator you specify the navigator and screen you want to navigate to. If your change your example first screen to this:

```
function First({navigation}) {
  return (
    <View style={{ flex: 1 }}>
      <Text>First</Text>
      <Button onPress={() => {
        navigation.navigate("secondTab", {screen: "subsecond"})
      }} title={"Navigate to subsecond"} />
    </View>
  )
}
```

You'll see that there is no way for you to go to the second screen from subsecond unless your Tab.Navigator is set with lazy=false. 

Maybe @satya164 or someone else from the maintainers can shed some light on how this is supposed to work, but the way it works now seems broken. 

The BottomTabView.tsx has the following code:
```
              if (lazy && !loaded.includes(index) && !isFocused) {
                // Don't render a screen if we've never navigated to it
                return null;
              }
```

So basically if lazy is true we never render the Stack.Navigator inside that tab. Since we don't render it - there is no initial screen. When navigate manually or from deep link the state seems to be such that it omits the initial route.

**Comment by Garkavenko** [comment]
@compojoom 
I agree
and i think, issue with bottom tabs is just a special case
there will be many other issues related to Stack.Navigator's initialRouteName prop, until we don't have some way to tune default screen, which will be rendered always as first screen in stack

hope, Maintainers will pay attention to the problem

**Comment by N1ghtly** [comment]
Just ran into this issue myself, can confirm that `lazy={false}` on the tav navigator fixed it for me.

**Comment by Garkavenko** [comment]
@N1ghtly but it's not a solution for me, as i described above

**Comment by lucasmotta** [comment]
Facing a simular issue, but setting `lazy={false}` does not work for me either. I've added some logging on the `useLinking.native.ts` to see what's happening to the links:

This is the state coming from `getStateFromPathRef`:
```json
{
  "routes": [
    {
      "name": "AppNavigator",
      "state": {
        "routes": [
          {
            "name": "App",
            "state": {
              "routes": [
                {
                  "name": "LearnNavigator",
                  "state": {
                    "index": 1,
                    "routes": [
                      {
                        "name": "Learn"
                      },
                      {
                        "name": "LearnArticle",
                        "params": {
                          "articleId": "15"
                        }
                      }
                    ]
                  }
                }
              ]
            }
          }
        ]
      }
    }
  ]
}
```

This is the action that's coming from `getActionFromState`:
```json
{
  "type": "NAVIGATE",
  "payload": {
    "name": "AppNavigator",
    "params": {
      "screen": "App",
      "params": {
        "screen": "LearnNavigator",
        "params": {
          "screen": "LearnArticle",
          "params": {
            "articleId": "15"
          }
        }
      }
    }
  }
}
```

So on the action itself there's nothing related to the `Learn` screen, that it should be the `initialRoute`.  So for testing purposes, if I dispatch a reset action using the CommonActions with the state, it works fine:

```js
navigation.dispatch(
  CommonActions.reset({
    routes: [
      {
        name: 'App',
        state: {
          routes: [
            {
              name: 'LearnNavigator',
              state: {
                index: 1,
                routes: [
                  {
                    name: 'Learn',
                  },
                  {
                    name: 'LearnArticle',
                    params: {
                      articleId: '15',
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    ],
  })
)
``` 

So I wonder if the issue is with the `getActionFromState`? Can we use the `CommonActions.reset(state)` instead?

**Comment by Garkavenko** [comment]
@lucasmotta yep, `CommonActions.reset(state)` will work, but if you have a complex navigation tree, it will take a long time for operation.
For example, i tried that in my real project (in dev mode) and it was taking about 4-5 seconds. It's extremely slow even for dev mode

**Comment by lucasmotta** [comment]
@Garkavenko yeah I noticed that on my app as well - even thou it's not ideal, but it's better than not working at all. So I'm using `patch-package` package for the time being with this fix.

**Comment by aiba** [comment]
I'm also encountering this issue (disappearing back button when useLinking goes to a screen deep in a stack nav).

lazy=false on the stack nav didn't work for me.

@lucasmotta - could you explain the `CommonsActions.reset(state)` workaround?  Where in the code would this go?  Could I also use `patch-package` to share your fix?

Thanks!

**Comment by IronTony** [comment]
Any real solution and an example? Thank you in advance

**Comment by aiba** [comment]
I couldn't get react-navigation's useLinking to work, so what I did instead was:

* Wrote a function, `navigateToPath`, that takes a string like "/screen-a/screen-b?foo=1/screen-c?bar=2" and parses it into a structure you can use to call `rootNavRef.navigate(...)` (as documented [here](https://reactnavigation.org/docs/navigating-without-navigation-prop)).
* Wrote my own hook, `useUrlLinking`, that calls plain react-native's `Linking.getInitialURL` and `Linking.addEventListener`.  Both of these handler functions call `navigateToPath`.
* Called `useUrlLinking` from the component containing the root `NavigationContainer`.

It wasn't so bad, and actually having the function `navigateToPath` turned out to be really useful elsewhere in the codebase, e.g. a Button `onPress: () => { navigateToPath("/root/home-tabs/settings/account"); }`.

**Comment by IronTony** [comment]
@aiba Thank you for the explanation. Do you have a gist or a repo with some code?

**Comment by aiba** [comment]
Happy to share, though I'm writing in ClojureScript.  Hopefully this is somewhat useful anyway.

https://gist.github.com/aiba/209243d709eb2cde953f31e0390ec330

**Comment by IronTony** [comment]
Thank you. I will try to understand it and convert to TS/JS

**Comment by satya164** [comment]
https://reactnavigation.org/docs/configuring-links/#rendering-an-initial-route

**Comment by IronTony** [comment]
@satya164 I tried to follow it but I don't understand how to "merge" with https://reactnavigation.org/docs/deep-linking/ and apply it with a tabbar and multiple stacknavigators

**Comment by lucasmotta** [comment]
Hey @satya164 we did follow the guides, but if you see the comments above you can see  it has some bugs that are not resolved - so I think this ticket is still valid.

I can create another one that's more specific to the issue I encountered above, what would you prefer?

**Comment by satya164** [comment]
Please update to latest version first before opening new issues.

**Comment by IronTony** [comment]
@satya164 ok, I will. Thank you. Is it just a hard request to have a deeper documentation on nested stack (stack navigator and bottom tabs) for deep linking?

All the best

**Comment by satya164** [comment]
what is the issue with stack navigator and bottom tabs? there isn't anything special to do for them

**Comment by IronTony** [comment]
@satya164 I have updated to the latest react-navigation packages, but I really don't understand how can I navigate to a nested route.
Here is my situation:

RootStack:
```
<RootStack.Navigator>
    <RootStack.Screen name="Main" component={TabsScreen} />
    <RootStack.Screen name="SingleNews" component={SinglePost} />
</RootStack.Navigator>
```

TabsScreen:
```
<BottomTabs.Navigator initialRouteName="Page1">
    <BottomTabs.Screen name="Page1" component={Page1Stack} />
    <BottomTabs.Screen name="Page2" component={Page2Stack} />
</BottomTabs.Navigator>
```

Page1Stack:
```
<Page1Stack.Navigator initialRouteName="Home">
    <Page1Stack.Screen name="Home" component={Homepage} />
    <Page1Stack.Screen name="InnerPage" component={InnerPage} />
    <Page1Stack.Screen name="AllNews" component={AllNews} />
</Page1Stack.Navigator>
```

App.tsx:
```
const config = {
  Main: {
    initialRouteName: 'Page1',
    screens: {
      Page1: 'pageone',
      Page2: 'pagetwo',
    },
  },
  SingleNews: 'singlenews/:id'
};

const [isReady, setIsReady] = useState(false);
const [initialState, setInitialState] = useState();

 const { getInitialState } = useLinking(navigationRef, {
    prefixes: ['https://my.app.us', 'mycustomapp://'],
    config,
  });

<NavigationContainer ref={navigationRef} initialState={initialState}>
.....
</NavigationContainer>

Using this line as DOC says for my emulator iOS:
`xcrun simctl openurl booted mycustomapp://pageone`
```

I thought to navigate to that page, but it seems not.
Maybe I didn't understand the DOC, sorry.

**Comment by IronTony** [comment]
@satya164 was my fault in the name I pass to `xcrun simctl openurl booted`
Working now thx

**Comment by vetrivendhan48** [comment]
Hi @Garkavenko , 

I actually found the issue replication scenerio,
This is due to absence of 'backBehavior' property.
I am expecting a proper fix for that.

To replicate the issue use the following sample code.
```
   <Tab.Navigator  backBehavior={"history"}
      // lazy={false}
    >
```

@IronTony , @satya164 - Workaround is to change ` backBehavior={"initialRoute"}` .

Thanks guys.

**Comment by syedamirali14** [comment]
@vetrivendhan48 
this is not working for me :(

**Comment by GustavoBonfimS** [comment]
I was using the lib in version 5.2 and it wasn't working. To not have Breaking changes I updated to the latest version of 5.x and everything worked as expected.

```
yarn add @react-navigation/native@^5.x
```

```
yarn add @react-navigation/stack@^5.x
```

**Comment by DeepBlueDevOps** [comment]
> https://reactnavigation.org/docs/configuring-links/#rendering-an-initial-route

I was having a problem where deep linking was working on initial linked navigation. However, each time a link (e.g. App Invite or App share) the other screens in the stack would not appear. Setting the initialRouteName fixed the issue and other screens were accessible.

**Comment by TSoli** [comment]
https://github.com/react-navigation/react-navigation/issues/7698#issuecomment-619535053

This works without having to use `lazy={false}`.
